pyfarm/agent/utility.py

```diff
@@ -392,3 +393,51 @@ def remove_file(
             raise
     else:
         logger.info("Removed %s", path)
+
+
+def remove_directory(
+        path, retry_on_exit=False, raise_=True, ignored_errnos=(ENOENT, )):
+    """
+    Simple function to **recursively** remove the provided directory or retry on
+    exit if requested.  This function standardizes the log output, ensures
+    it's only called once per path on exit and handles platform
+    specific exceptions (ie. ``WindowsError``).
+
+    :param bool retry_on_exit:
+        If True, retry removal of the file when Python exists.
+
+    :param bool raise_:
+        If True, raise an exceptions produced.  This will always be
+        False if :func:`remove_file` is being executed by :module:`atexit`
+
+    :param tuple ignored_errnos:
+        A tuple of ignored error numbers.  By default this function
+        only ignores ENOENT.
+    """
+    try:
+        shutil.rmtree(path)
+    except (WindowsError, OSError, IOError) as error:
+        if error.errno in ignored_errnos:
+            logger.debug(
+                "Failed to remove %s (%s)", path, errorcode[error.errno])
+            return
+
+        if retry_on_exit:
+            logger.warning(
+                "Failed to remove %s (%s), will retry on exit",
+                path, errorcode[error.errno])
+
+            # Make sure we're not added multiple times
```

Comment:
Potentially over time yes because you could add the same path over and over by accident or in a loop.  The agent can run for quite a while so I'm trying to make sure we're not unnecessarily building up entries in a list that will persist until shutdown (which will also make the shutdown take longer).